File: vitaly.aksenov/HW01/src/Image.java
import java.io.DataInputStream;
import java.io.IOException;

public class Image {
	int n, m;
	byte[][] b;
	byte[] vector;

	public Image(int n, int m) {
		this.n = n;
		this.m = m;
		b = new byte[n][m];
		vector = new byte[n * m];
	}

	public void read(DataInputStream in) throws IOException {
		in.read(vector);
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				b[i][j] = vector[i * m + j];
			}
		}
	}

	public int length() {
		return n * m;
	}
}

File: arkady.rost/src/main/scala/ru/ifmo/ctd/rost/ml/Image.scala
package ru.ifmo.ctd.rost.ml

import java.io.InputStream

class Image private(val rows: Int, val columns: Int, val data: Array[Int])

object Image {
  abstract class TRUE
  abstract class FALSE

  def newBuilder(rows : Int, columns : Int) = ImageBuilder.empty(rows, columns)

  class ImageBuilder[HD] private(val n: Int, val m: Int, val data : Option[Array[Int]]) {

    def fromStream(in: InputStream) = new ImageBuilder[TRUE](n, m, Some(Stream.continually(in.read).take(n * m).toArray))
  }

  implicit def enableBuild(builder : ImageBuilder[TRUE]) = new {
    def build() =  new Image(builder.n, builder.m, builder.data.get)
  }

  object ImageBuilder {
    def empty(n : Int, m : Int) = {
      assert(n > 0)
      assert(m > 0)
      new ImageBuilder[FALSE](n, m, None)
    }
  }
}
